### PR TITLE
Stress testing with Artillery

### PR DESCRIPTION
**Pull request by sjamcox**
## Trello Ticket
https://trello.com/c/Iqa6kx41

## Purpose of Pull Request
Created config file and middleware for stress testing.
Three scenarios for stress testing: Visit 1 page, Visit 5 pages, Visit 1 page and leave a review

**Note:** With the way the client is set up, each page load makes 4 requests to the reviews service. You'll see that reflected in the stress testing. I recognize this isn't optimal, but without overhauling the client (which I think is out of bounds) that's how it is.

Some minor file cleanup is also included.

### CODE REVIEW

File: Tests/processor.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    context.vars.wait = Math.floor(Math.random() * 20) + 3;
+    return next();
+  },
+
+  generatePageCount: (context, ee, next) => {
+    context.vars.pages = Math.floor(Math.random() * 5) + 1;
+    return next();
+  },
+
+  generateProductReview: (requestParams, context, ee, next) => {
```
Comment (tyler-thee-creator) [inline Tests/processor.js:20]: Was trying to figure out how to add json for a post, this clears it up, nicely done

File: Tests/service-test.yml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+      arrivalRate: 75
+      rampTo: 100
+    - duration: 300
+      arrivalRate: 100
+      rampTo: 200
+  processor: "./processor.js"
+
+scenarios:
+  - name: "View product page"
+    weight: 15
```
Comment (tyler-thee-creator) [inline Tests/service-test.yml:22]: DIdn't know about a weight variable for scenarios but just looked it up - pretty cool

File: Tests/service-test.yml [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    - duration: 60
+      arrivalRate: 10
+      rampTo: 50
+    - duration: 60
+      arrivalRate: 50
+      rampTo: 75
+    - duration: 60
+      arrivalRate: 75
+      rampTo: 100
+    - duration: 300
```
Comment (tyler-thee-creator) [inline Tests/service-test.yml:15]: Is there a reason your duration here is 300 and not 60 like the others? Also is there a reason you're not trying 1000 RPS? Maybe 1000 was too much for you system so you downgraded to go up to 200?

File: Tests/processor.js [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        'http://placeimg.com/480/640/tech',
+      ],
+      productId: faker.random.number({ min: 1, max: 10000000 }),
+      customerId: faker.random.number({ min: 1, max: 10000000 })
+    };
+    return next();
+  }
+
+};
+
```
Comment (tyler-thee-creator) [inline Tests/processor.js:48]: Great looking processor file

File: Tests/service-test.yml [hunk 5]
```diff
@@ -0,0 +1,65 @@
+config:
+  target: "http://localhost:4006"
+  http:
+    timeout: 10
```
Comment (tyler-thee-creator) [inline Tests/service-test.yml:4]: Also didn't know about the timeout variable for your config - also pretty cool, very nice

File: database-postgres/dbHelpers.js [hunk 6]
```diff
@@ -57,7 +57,7 @@ exports.getReview = (id) => {
 exports.createReview = (body) => {
   body.reviewId = faker.random.uuid();
   const keys = Object.keys(body)
-  const columns = keys.join(',')
+  const columns = keys.map(key => `"${key}"`).join(', ')
```
Comment (tyler-thee-creator) [inline database-postgres/dbHelpers.js:60]: I like this line here, super simple but gets the job done

**Review by tyler-thee-creator — APPROVED** [review]
Looks really good! Nicely done

File: Tests/service-test.yml (hunk 3 above)
Comment (sjamcox) [reply Tests/service-test.yml:15]: For the duration question, I wanted to slowly ramp up from 100 to 200 to better see if it was the sustained load or actual stress that was causing the high cpu usage. I could have continued to increase smaller amounts in 60 second increments, but slowly increasing over 5 minutes seemed like less code for the same effect.

For the requests question, each scenario actually launches at least 4 requests, so 200 users doing one page view is actually 800 requests. I also have the scenario for one user to do five page views over a period of time, so all in all they'd launch 20 requests each. It worked out to where I hit a peak of nearly 1600 rps, which came back down to a consistent 1300 rps after I bounced off that ceiling.

**Pull request merged**